pkg/protocol/engines.go

```diff
@@ -0,0 +1,261 @@
+func (e *Engines) CleanupCandidates() error {
+	activeDir := filepath.Base(e.Main.Get().Storage.Directory())
+
+	dirs, err := e.directory.SubDirs()
+	if err != nil {
+		return ierrors.Wrapf(err, "unable to list subdirectories of %s", e.directory.Path())
+	}
+	for _, dir := range dirs {
+		if dir == activeDir {
+			continue
+		}
+		if err := e.directory.RemoveSubdir(dir); err != nil {
+			return ierrors.Wrapf(err, "unable to remove subdirectory %s", dir)
+		}
+	}
+
+	return nil
+}
+
+// infoFilePath returns the path to the engine info file.
+func (e *Engines) infoFilePath() string {
+	return e.directory.Path(engineInfoFile)
+}
+
+// loadEngineInstanceFromSnapshot loads an engine instance from a snapshot.
+func (e *Engines) loadEngineInstanceFromSnapshot(engineAlias string, snapshotPath string) *engine.Engine {
+	errorHandler := func(err error) {
+		e.protocol.LogError("engine error", "err", err, "name", engineAlias[0:8])
+	}
+
+	e.protocol.Options.EngineOptions = append(e.protocol.Options.EngineOptions, engine.WithSnapshotPath(snapshotPath))
+
+	return e.loadEngineInstanceWithStorage(engineAlias, storage.Create(e.directory.Path(engineAlias), DatabaseVersion, errorHandler, e.protocol.Options.StorageOptions...))
+}
+
+// loadEngineInstanceWithStorage loads an engine instance with the given storage.
+func (e *Engines) loadEngineInstanceWithStorage(engineAlias string, storage *storage.Storage) *engine.Engine {
+	return engine.New(e.protocol.Logger, e.protocol.Workers.CreateGroup(engineAlias), storage, e.protocol.Options.PreSolidFilterProvider, e.protocol.Options.PostSolidFilterProvider, e.protocol.Options.BlockDAGProvider, e.protocol.Options.BookerProvider, e.protocol.Options.ClockProvider, e.protocol.Options.BlockGadgetProvider, e.protocol.Options.SlotGadgetProvider, e.protocol.Options.SybilProtectionProvider, e.protocol.Options.NotarizationProvider, e.protocol.Options.AttestationProvider, e.protocol.Options.LedgerProvider, e.protocol.Options.SchedulerProvider, e.protocol.Options.TipManagerProvider, e.protocol.Options.TipSelectionProvider, e.protocol.Options.RetainerProvider, e.protocol.Options.UpgradeOrchestratorProvider, e.protocol.Options.SyncManagerProvider, e.protocol.Options.EngineOptions...)
+}
+
+// syncMainEngineFromMainChain syncs the main engine from the main chain.
+func (e *Engines) syncMainEngineFromMainChain() (shutdown func()) {
+	return e.protocol.Chains.Main.WithNonEmptyValue(func(mainChain *Chain) (shutdown func()) {
+		return e.Main.DeriveValueFrom(reactive.NewDerivedVariable(func(currentMainEngine *engine.Engine, newMainEngine *engine.Engine) *engine.Engine {
+			return lo.Cond(newMainEngine == nil, currentMainEngine, newMainEngine)
+		}, mainChain.Engine))
+	})
+}
+
+// syncMainEngineInfoFile syncs the engine info file with the main engine.
+func (e *Engines) syncMainEngineInfoFile() (shutdown func()) {
+	return e.Main.OnUpdate(func(_ *engine.Engine, mainEngine *engine.Engine) {
```

Comment:
Why not `e.Main.WithNonEmptyValue` ?

Reply:
because we do not want to setup a reactive behavior with a shutdown function, but we just want to setup a simple function call.